API/Controllers/UsersController.cs

```diff
@@ -1,37 +1,103 @@
 ﻿using API.Models;
+using API.Models.AccountModels;
 using API.Services;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Identity;
 using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Configuration;
+using System.Threading.Tasks;
 
 namespace API.Controllers
 {
     [Route("api/v1/[controller]/[action]")]
     [Produces(ApiContentType)]
     [ApiController]
+    [Authorize]
     public class UsersController : ApiControllerBase
     {
-        public UsersController(ApiContext context, Iserializer serializer, IErrorFormatter errorFormatter) :
-            base(context, serializer, errorFormatter) { }
+        private readonly UserManager<UserDTO> _userManager;
+        private readonly SignInManager<UserDTO> _signInManager;
+        private readonly JwtService _jwt;
+        public UsersController(
+            ApiContext context,
+            Iserializer serializer,
+            IErrorFormatter errorFormatter,
+            IConfiguration config,
+            UserManager<UserDTO> userManager,
+            SignInManager<UserDTO> signInManager) :
+            base(context, serializer, errorFormatter)
+        {
+            _userManager = userManager;
+            _signInManager = signInManager;
+            _jwt = new JwtService(config);
+        }
 
         [HttpPost]
-        public IActionResult Signup(UserDTO user)
+        [AllowAnonymous]
+        public async Task<IActionResult> Signup(RegisterModel model)
         {
             if (!IsValidApiRequest())
             {
                 return ApiBadRequest("Invalid headers!");
             }
 
-            return NotImplementedError("Signup Not Implemented!");
+            if (ModelState.IsValid)
+            {
+                var user = new UserDTO { UserName = model.Username ?? model.Email, Email = model.Email };
+                var result = await _userManager.CreateAsync(user, model.Password);
+                if (result.Succeeded)
+                {
+                    return Created("", new
+                    {
+                        token = _jwt.GenerateSecurityToken(new User()
+                        {
+                            Username = user.UserName,
+                            Email = user.Email
+                        })
+                    });
+                }
+
+                return BadRequest(result);
```

Comment:
After testing with invalid credentials, I receive an error json like this:
```json
{
    "succeeded": false,
    "errors": [
        {
            "code": "PasswordTooShort",
            "description": "Passwords must be at least 6 characters."
        },
        {
            "code": "PasswordRequiresDigit",
            "description": "Passwords must have at least one digit ('0'-'9')."
        },
        {
            "code": "PasswordRequiresUpper",
            "description": "Passwords must have at least one uppercase ('A'-'Z')."
        }
    ]
}
```
Please correctly format error object as specified here: https://github.com/KamandaX/Wiki/blob/master/API_Contracts.md
Detailed validation check should be done at frontend.

Reply:
Same with this:
```json
{
    "succeeded": false,
    "errors": [
        {
            "code": "DuplicateUserName",
            "description": "User name 'admin' is already taken."
        }
    ]
}
```

Reply:
Fixed.